### PR TITLE
net-cpp: init at 3.1.0

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

Working towards #99090.

[net-cpp](https://gitlab.com/ubports/development/core/lib-cpp/net-cpp), a C++11 networking library. Needed by Lomiri's location service.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/ne/net-cpp/package.nix [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      hash = "sha256-nsGkZBuqahsg70PLUxn5EluDjmfZ0/wSnOYimfAI4ag=";
+    })
+  ];
+
+  postPatch = if finalAttrs.doCheck then ''
+    # Use wrapped python. Removing just the /usr/bin doesn't seem to work?
+    substituteInPlace tests/httpbin.h.in \
+      --replace '/usr/bin/python3' '${lib.getExe pythonEnv}'
+  '' else ''
+    sed -i CMakeLists.txt -e '/add_subdirectory(tests)/d'
```
Comment (pbsds) [inline pkgs/by-name/ne/net-cpp/package.nix:64]:
```suggestion
    substituteInPlace CMakeLists.txt --replace "add_subdirectory(tests)" ""
```

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 260869` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>net-cpp</li>
    <li>net-cpp.dev</li>
    <li>net-cpp.doc</li>
  </ul>
</details>

LGTM, just a nit

File: pkgs/by-name/ne/net-cpp/package.nix (hunk 1 above)
Comment (OPNA2608) [reply pkgs/by-name/ne/net-cpp/package.nix:64]: Personal preference is `substituteInPlace` when I actually want to just make a change, and `sed` with `/d` or `/a new thing here` for deletions & additions respectively.

Either way, I'll prolly just upstream an option to make the tests optional and `fetchpatch` it instead, i.e. https://gitlab.com/ubports/development/core/lomiri-download-manager/-/merge_requests/23. Just need to find the time to actually do it.

Comment (OPNA2608) [reply pkgs/by-name/ne/net-cpp/package.nix]: I've submitted the patch & `fetchpatch` it now. Reason for `substituteInPlace` has been noted, I'll try to keep it in mind.

File: pkgs/by-name/ne/net-cpp/package.nix [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    })
+
+    # Be more lenient with how quickly HTTP test server must be up, for slower hardware / archs
+    (fetchpatch {
+      url = "https://salsa.debian.org/ubports-team/net-cpp/-/raw/941d9eceaa66a06eabb1eb79554548b47d4a60ab/debian/patches/1007_wait-for-flask.patch";
+      hash = "sha256-nsGkZBuqahsg70PLUxn5EluDjmfZ0/wSnOYimfAI4ag=";
+    })
+  ];
+
+  postPatch = lib.optionalString finalAttrs.doCheck ''
```
Comment (pbsds) [inline pkgs/by-name/ne/net-cpp/package.nix:67]:
```suggestion
  postPatch = lib.optionalString finalAttrs.finalPackage.doCheck ''
```

This covers the case where `doCheck = false;` is set using an override. I remembed the cross people discussing it on matrix.

**Review by pbsds — APPROVED** [review]

File: pkgs/by-name/ne/net-cpp/package.nix (hunk 2 above)
Comment (pbsds) [reply pkgs/by-name/ne/net-cpp/package.nix:67]: Quick to do treewide however

**Pull request merged**